### PR TITLE
122 change n cores to n_vcpus and allow user to send number of vcpus

### PR DESCRIPTION
**Pull request by Pbarbosa92**
In this pr we we give the user more control over the simulations.

We now allow the user to send as a parameter to some simulations the number of cpus that he wants to run (n_vcpu).
Because some machines might have (or not) hyperthreading, we also allow the user to send a flag that will define if we use hyperthreading or not (use_threads_cpus - True by default).

With this the user can have a machine with 10 cores (20 threads) and decide if he wants his simulation:
- To run on 10 physical cores (n_vcpus=10, use_threads_cpus=False)
- To run on 10 threads (n_vcpus=10, use_threads_cpus=True)
- On all available threads (n_vcpus=20, use_threads_cpus=True)

**Possible problem**
The user can ask for 15 threads and ask for threading off, and it will get an error like this
"There are not enough slots available in the system to satisfy the 19
slots that were requested by the application:"

The problem is that this error only shows on the stderr in the task output files. The user will only see that the task failed.

We can let the error do the mistake **OR** create a back-end endpoint that tells us the number of cores for a machine type and we can generate an exception. 
The big issue with this is that the task can go to a queue for 2 hours, and the user will only have the task failed after 2 hours of waiting.

**Task** [here](https://github.com/inductiva/tasks/issues/122)

### CODE REVIEW

File: inductiva/utils/meta.py [hunk 1]
```diff
@@ -59,3 +60,32 @@ def get_method_name(function_ptr) -> str:
     module_name = ".".join(module_name_useful_segments)
 
     return f"{module_name}.{function_ptr.__name__}"
+
+
+# pylint: disable=unused-argument
+def custom_format_warning(msg, *args, **kwargs):
+    """Custom format for warnings."""
+    return str(msg) + "\n"
```
Comment (sergiomsantos) [inline inductiva/utils/meta.py:68]: why is this necessary?

File: inductiva/utils/meta.py [hunk 2]
```diff
@@ -59,3 +60,32 @@ def get_method_name(function_ptr) -> str:
...
+
+
+def deprecated_arg(**kwargs_decorator):
```
Comment (sergiomsantos) [inline inductiva/utils/meta.py:71]: doc string?

File: inductiva/utils/meta.py [hunk 3]
```diff
@@ -59,3 +60,32 @@ def get_method_name(function_ptr) -> str:
...
+
+    def wrapper(func):
+
+        def inner_wrapper(*args, **kwargs):
+            decorator_keys = kwargs_decorator.keys()
+            function_keys = kwargs.keys()
+
+            warnings.formatwarning = custom_format_warning
```
Comment (sergiomsantos) [inline inductiva/utils/meta.py:79]: why is this necessary?

File: inductiva/simulators/xbeach.py [hunk 4]
```diff
@@ -12,19 +13,21 @@ def __init__(self):
...
+            on: Optional[types.ComputationalResources] = None,
+            storage_dir: Optional[types.Path] = "",
+            extra_metadata: Optional[dict] = None,
+            **kwargs) -> tasks.Task:
         """Run the simulation.
 
         Args:
             input_dir: Path to the directory of the simulation input files.
             sim_config_filename: Name of the simulation configuration file.
+            n_vcpus: Number of vCPUs (all by default) to use for the simulation.
```
Comment (sergiomsantos) [inline inductiva/simulators/xbeach.py:30]:
```suggestion
            n_vcpus: Number of vCPUs to use in the simulation. If not provided (default), all vCPUs will be used.
```

File: inductiva/utils/meta.py [hunk 5]
```diff
@@ -59,3 +60,32 @@ def get_method_name(function_ptr) -> str:
...
+
+            for decorator_key in decorator_keys:
+                if decorator_key in function_keys:
+                    warnings.warn(f"{decorator_key} is deprecated, use "
+                                  f"{kwargs_decorator[decorator_key]} instead.")
+                    kwargs[
+                        kwargs_decorator[decorator_key]] = kwargs[decorator_key]
```
Comment (sergiomsantos) [inline inductiva/utils/meta.py:86]: this can be made cleaner: no need to get values multiple times from the dict.
Also, to get the common keys you can use `set( kwargs_decorator) & set(kwargs)`

File: inductiva/simulators/fds.py [hunk 6]
```diff
@@ -12,21 +13,22 @@ def __init__(self):
...
+            on: Optional[types.ComputationalResources] = None,
+            storage_dir: Optional[types.Path] = "",
+            extra_metadata: Optional[dict] = None,
+            **kwargs) -> tasks.Task:
         """Run the simulation.
 
         Args:
             input_dir: Path to the directory of the simulation input files.
             sim_config_filename: Name of the simulation configuration file.
+            n_vcpus: Number of vCPUs (all by default) to use for the simulation.
```
Comment (sergiomsantos) [inline inductiva/simulators/fds.py:31]: see my other comment

File: inductiva/utils/meta.py (hunk 1 above)
Comment (Pbarbosa92) [reply inductiva/utils/meta.py:68]: Without this the warning shows like this:
```
/Users/paulobarbosa/Documents/Gits/inductiva/inductiva/utils/meta.py:83): UserWarning: n_cores is deprecated, use n_vcpus instead.
  warnings.warn(f"{decorator_key} is deprecated, use "
```

Since meta.py is the code of the decorator that is meaningless to the user.

File: inductiva/simulators/xbeach.py [hunk 7]
```diff
@@ -12,19 +13,26 @@ def __init__(self):
...
-        sim_config_filename: Optional[str] = "params.txt",
-        on: Optional[types.ComputationalResources] = None,
-        storage_dir: Optional[types.Path] = "",
-        extra_metadata: Optional[dict] = None,
-    ) -> tasks.Task:
+    @meta.deprecated_arg(n_cores="n_vcpus")
+    def run(self,
+            input_dir: types.Path,
+            n_vcpus: Optional[int] = None,
+            use_hwthread: bool = None,
```
Comment (IvanPombo) [inline inductiva/simulators/xbeach.py:20]: The default value of `use_hwthread` should be `True` everywhere. Let's assume users want to use all vcpus they have available.

File: inductiva/utils/meta.py [hunk 8]
```diff
@@ -59,3 +60,35 @@ def get_method_name(function_ptr) -> str:
...
+        def inner_wrapper(*args, **kwargs):
+
+            for key in set(deprecated) & set(kwargs):
+                new_key = deprecated[key]
+                value = kwargs.pop(key)
+                kwargs[new_key] = value
+
+                warnings.warn(
+                    f"{key} is deprecated, use "
+                    f"{new_key} instead.",
```
Comment (IvanPombo) [inline inductiva/utils/meta.py:87]:
```suggestion
                    f"The {key} argument was deprecated, and substituted with {new_key}.",
```

**Review by IvanPombo — COMMENTED** [review]
Almost ready to go!

File: inductiva/simulators/openfoam.py [hunk 9]
```diff
@@ -23,19 +24,26 @@ def __init__(self, version: str = "foundation"):
...
 
-    def run(
-        self,
-        input_dir: types.Path,
-        commands: types.Commands,
-        on: Optional[types.ComputationalResources] = None,
-        storage_dir: Optional[types.Path] = "",
-        extra_metadata: Optional[dict] = None,
-    ) -> tasks.Task:
+    @meta.deprecated_arg(n_cores="n_vcpus")
```
Comment (IvanPombo) [inline inductiva/simulators/openfoam.py:27]: This decorator should not be here, since the argument didn't exist at all for this simulator.

File: inductiva/simulators/reef3d.py [hunk 10]
```diff
@@ -13,17 +14,24 @@ def __init__(self):
...
         self.api_method_name = "reef3d.reef3d.run_simulation"
 
-    def run(
-        self,
-        input_dir: types.Path,
-        on: Optional[types.ComputationalResources] = None,
-        storage_dir: Optional[types.Path] = "",
-        extra_metadata: Optional[dict] = None,
-    ) -> tasks.Task:
+    @meta.deprecated_arg(n_cores="n_vcpus")
```
Comment (IvanPombo) [inline inductiva/simulators/reef3d.py:17]: Same as above.

File: inductiva/simulators/swash.py [hunk 11]
```diff
@@ -12,19 +13,26 @@ def __init__(self):
...
 
-    def run(
-        self,
-        input_dir: types.Path,
-        sim_config_filename: str,
-        on: Optional[types.ComputationalResources] = None,
-        storage_dir: Optional[types.Path] = "",
-        extra_metadata: Optional[dict] = None,
-    ) -> tasks.Task:
+    @meta.deprecated_arg(n_cores="n_vcpus")
```
Comment (IvanPombo) [inline inductiva/simulators/swash.py:16]: here as well.

File: inductiva/simulators/xbeach.py [hunk 12]
```diff
@@ -12,19 +13,26 @@ def __init__(self):
...
 
-    def run(
-        self,
-        input_dir: types.Path,
-        sim_config_filename: Optional[str] = "params.txt",
-        on: Optional[types.ComputationalResources] = None,
-        storage_dir: Optional[types.Path] = "",
-        extra_metadata: Optional[dict] = None,
-    ) -> tasks.Task:
+    @meta.deprecated_arg(n_cores="n_vcpus")
```
Comment (IvanPombo) [inline inductiva/simulators/xbeach.py:16]: here as well

**Comment by Pbarbosa92** [comment]
**QA**

To be testes at the same time as https://github.com/inductiva/inductiva-web-api/pull/633

File: inductiva/utils/meta.py [hunk 13]
```diff
@@ -59,3 +60,35 @@ def get_method_name(function_ptr) -> str:
     module_name = ".".join(module_name_useful_segments)
 
     return f"{module_name}.{function_ptr.__name__}"
+
+
+# pylint: disable=unused-argument
+def custom_format_warning(msg, *args, **kwargs):
```
Comment (sergiomsantos) [inline inductiva/utils/meta.py:66]: is this still necessary?

Comment (Pbarbosa92) [reply inductiva/utils/meta.py]: Removed
